src/client.rs

```diff
@@ -52,28 +45,29 @@ impl ServerConnection {
     }
 
     /// Send `request` to the server, read and return a `ServerResponse`.
-    pub fn request(&mut self, request : ClientRequest)
-                   -> ProtobufResult<ServerResponse> {
+    pub fn request(&mut self, request: Request) -> Result<Response> {
         trace!("ServerConnection::request");
-        try!(request.write_length_delimited_to_writer(&mut self.writer));
-        try!(self.writer.flush().or_else(|e| Err(ProtobufError::IoError(e))));
+        let bytes = bincode::serialize(&request, bincode::Infinite)?;
+        let mut len = [0; 4];
+        BigEndian::write_u32(&mut len, bytes.len() as u32);
+        self.writer.write_all(&len)?;
+        self.writer.write_all(&bytes)?;
+        self.writer.flush()?;
         trace!("ServerConnection::request: sent request");
         self.read_one_response()
     }
 
     /// Read a single `ServerResponse` from the server.
-    pub fn read_one_response(&mut self) -> ProtobufResult<ServerResponse> {
+    pub fn read_one_response(&mut self) -> Result<Response> {
         trace!("ServerConnection::read_one_response");
-        //FIXME: wish `parse_length_delimited_from` worked here!
-        let len = try!({
-            let mut is = CodedInputStream::from_buffered_reader(&mut self.reader);
-            is.read_raw_varint32()
-        });
+        let mut bytes = [0; 4];
+        self.reader.read_exact(&mut bytes)?;
```

Comment:
Would things look nicer overall if we used tokio here instead of blocking IO? Originally I used blocking IO in the client because the mio stuff was so complicated.